### PR TITLE
Add up_squared_sbl configuration to work zephyr with SBL

### PR DESCRIPTION
**Pull request by myguitar**
This adds a primitive board configuration for the UP Squared board to work properly with SBL(Slim Bootloader).
- https://github.com/slimbootloader/slimbootloader
- https://slimbootloader.github.io/

The BOARD shall be 'up_squared_sbl' when building applications.
  cmake -GNinja -DBOARD=up_squared_sbl ..
  ninja

### CODE REVIEW

File: cmake/dts.cmake [hunk 1]
```diff
@@ -10,6 +10,9 @@
 # See ~/zephyr/doc/dts
 set(GENERATED_DTS_BOARD_H    ${PROJECT_BINARY_DIR}/include/generated/generated_dts_board.h)
 set(GENERATED_DTS_BOARD_CONF ${PROJECT_BINARY_DIR}/include/generated/generated_dts_board.conf)
+
+include(${BOARD_DIR}/dts.cmake OPTIONAL)
+
```
Comment (SebastianBoe) [inline cmake/dts.cmake]: This should be in a dedicated PR.

**Review by SebastianBoe — CHANGES_REQUESTED** [review]

File: cmake/dts.cmake (hunk 1 above)
Comment (myguitar) [reply cmake/dts.cmake]: This PR has 2 separate commits. One of them is only for this change. Still need separate PR for this one?

Comment (nashif) [reply cmake/dts.cmake]: No, I do not think we need another PR for this as long as it is split into different commits. Thanks.

**Review by nashif — APPROVED** [review]

**Comment by nashif** [comment]
@SebastianBoe requiring multiple PRs for something like that is an overkill and complicates the contribution process. Not sure what is the rationale behind that. If you submit the dts change in another PR, it is very difficult to know the intention and the context, it the changes depending on it will not build and dependency will be difficult to track. Asking for different PRs is justified only when we really have unrelated contexts and changes.

**Comment by SebastianBoe** [comment]
The board CMake hook is not necessary and should be removed.

We want to avoid hooks because they cause divergence. When you can do anything,
everyone will find different and insignificantly different solutions to the same problem.

In this case, the problem is that the BOARD_DIR has several board's, and we need the
fixup file to be specific to the board, not the BOARD_DIR as has been convention.

There are two solutions to this problem. Either use the C preprocessor to distinguish between
the two fixup files instead of using CMake code (which is generally, and AFAICT in this case, desired) as shown in:
https://github.com/SebastianBoe/zephyr/commit/0c2968a0166b47995392cf3a9a43d0e7a55e2e49

Or change the dts_fixup.h detection logic such that the fixup file is board-dependent instead of board_dir-dependent (as we do with the .yaml, .dts, and _defconfig files).

**Comment by myguitar** [comment]
@nashif Thanks for your comment and your understanding the difficulty of contribution process.

@SebastianBoe Thanks for your comment and I totally understand your concern to use hooks. The hooks always have pros and cons.

Let me ask some questions. There are some override variables in cmake/dts.cmake which are DTS_SOURCE, DTS_COMMON_OVERLAYS, DTS_APP_BINDINGS, DTS_BOARD_FIXUP_FILE and DTS_SOC_FIXUP_FILE. How can we override those variables if required?

In addition, there is a user guide about dts override in https://docs.zephyrproject.org/latest/application/application.html?highlight=dts_source.
- This location can be overridden by setting the DTS_SOURCE CMake variable.
- The common overlays can be overridden by setting the DTS_COMMON_OVERLAYS CMake variable.
But, there is no CMake for a board specific one. We have board.cmake, but it's later than dts.cmake.

Your solution requires CONFIG_SBL_FIXUP=y in every applications' prj.conf. This should be manually done by developers. This may cause more divergence. What do you think?

**Comment by SebastianBoe** [comment]
> How can we override those variables if required?

From the application build script or from the command line.

>Your solution requires CONFIG_SBL_FIXUP=y in every applications' prj.conf.

We can solve that by having the board select SBL_FIXUP through Kconfig.

**Comment by myguitar** [comment]
@SebastianBoe That was why I added dts.cmake hook to avoid changing application build script or adding command line if dts variables need to be overridden. I will follow the same way without hook. According to your guideline, I will remove dts.cmake change and make changes only in up_squared directory.

**Review by SebastianBoe — APPROVED** [review]

**Comment by myguitar** [comment]
@nashif If you have different concern on 2nd changes, please let me know. I will follow up. Thanks.

**Comment by nashif** [comment]
@myguitar can you explain why we have two DTS files with different registers based on the bootloader being used? I am not following why is this needed.

**Comment by myguitar** [comment]
@nashif Unlikely ARM which has fixed MMIO in AMBA, x86 PCI devices can be mapped to any memory space through PCI enumeration in boot firmware. The original up_squared dts has MMIO addresses mapped from BIOS, but SBL is using different MMIO space for PCI device enumeration. If a user uses coreboot as a boot firmware for Zephyr, it will also require another dts for coreboot MMIO space. In fact, even if a user uses the same BIOS, the MMIO base in dts sometimes needs to be updated. It's because PCI devices will be mapped to different address if a user disables a PCI device in BIOS menu.
Therefore, a better way to get MMIO BAR of each devices is to read it from PCI config space directly in run-time instead of hard-coded values in dts. But, this change is a simple way to give SBL MMIO mapping to up_squared board when SBL is used.
